db/migrate/20190617191109_move_awx_credentials_to_authentications.rb

```diff
@@ -0,0 +1,143 @@
+  end
+
+  class MiqDatabase < ActiveRecord::Base; end
+
+  ENCRYPTED_ATTRIBUTES = %w[
+    auth_key
+    auth_key_password
+    become_password
+    password
+  ].freeze
+
+  FIELD_MAP = {
+    'authorize_password' => 'become_password',
+    'become_username'    => 'become_username',
+    'become_password'    => 'become_password',
+    'password'           => 'password',
+    'secret'             => 'auth_key',
+    'security_token'     => 'auth_key',
+    'ssh_key_data'       => 'auth_key',
+    'ssh_key_unlock'     => 'auth_key_password',
+    'username'           => 'userid',
+    'vault_password'     => 'password'
+  }.freeze
+
+  OPTIONS_FIELDS = %w[
+    authorize
+    become_method
+    client
+    domain
+    host
+    project
+    subscription
+    tenant
+  ].freeze
+
+  def up
+    embedded_ansible_authentications.each do |auth|
+      next unless auth.manager_ref
+
+      say_with_time("Migrating credential #{auth.name} from awx to vmdb") do
+        awx_info = awx_credential_info(auth.manager_ref)
+        update_authentication(auth, awx_info)
+      end
+    end
+  rescue PG::ConnectionBad
+    say("awx database is unreachable, credentials cannot be migrated")
+  end
+
+  private
+
+  def embedded_ansible_authentications
+    types = %w[
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::Credential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::CloudCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::MachineCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::NetworkCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::ScmCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::VaultCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::AmazonCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::AzureCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::GoogleCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::OpenstackCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::RhvCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::VmwareCredential
+    ]
+    Authentication.in_my_region.where(:type => types)
+  end
+
+  def awx_credential_info(awx_id)
+    cred_info = awx_connection.async_exec("SELECT inputs FROM main_credential WHERE id = $1::BIGINT", [awx_id]).first
+    # in case there is no matching credential on the awx side
+    return {} unless cred_info
+
+    JSON.parse(cred_info["inputs"])
+  end
+
+  def awx_connection
+    @awx_connection ||= PG::Connection.new(ApplicationRecord.connection.raw_connection.conninfo_hash.merge(:dbname => "awx").delete_blanks)
+  end
+
+  def update_authentication(auth, awx_info)
+    auth.options = auth.options.slice(*OPTIONS_FIELDS.map(&:to_sym)).presence
+
+    awx_info.each do |k, v|
+      if OPTIONS_FIELDS.include?(k)
+        auth.options ||= {}
```

Comment:
Confused by this line a bit I guess.  This should only need to run once, so is there is there a reason it isn't under line  87?

Also, seems like we should just be putting this guard in with the original `presence` line if it can possibly "turn `nil`"

```diff
- auth.options = auth.options.slice(*OPTIONS_FIELDS.map(&:to_sym)).presence
+ auth.options = auth.options.slice(*OPTIONS_FIELDS.map(&:to_sym)).presence || {}
```

Because there is no way you can do `auth.options.slice` if `auth.options == nil`, so my guess is that `nil` isn't a possible value for `auth.options` naturally, right?

Reply:
Actually I have a test case to change `options` to `nil` if we don't have any [[ref]](https://github.com/ManageIQ/manageiq-schema/pull/385/files#diff-b1b98e5c9ff8bdf8d7f905d42329aaceR91). And the column is nullable so I should take the nil case into account here.